clippy_lints/src/dereference.rs

```diff
@@ -631,27 +964,61 @@ fn report(cx: &LateContext<'_>, expr: &Expr<'_>, state: State, data: StateData)
                 app,
             );
         },
-        State::DerefedBorrow {
-            required_precedence,
-            msg,
-            ..
-        } => {
+        State::DerefedBorrow(state) => {
             let mut app = Applicability::MachineApplicable;
-            let snip = snippet_with_context(cx, expr.span, data.span.ctxt(), "..", &mut app).0;
+            let (snip, snip_is_macro) = snippet_with_context(cx, expr.span, span.ctxt(), "..", &mut app);
             span_lint_and_sugg(
                 cx,
                 NEEDLESS_BORROW,
-                data.span,
-                msg,
+                span,
+                state.msg,
                 "change this to",
-                if required_precedence > expr.precedence().order() && !has_enclosing_paren(&snip) {
+                if !snip_is_macro && expr.precedence().order() < pos.precedence() && !has_enclosing_paren(&snip) {
```

Comment:
Can this be done with `utils::Sugg`?

Reply:
`Sugg` doesn't take into account the snipped expression, nor does it take into account the precedence of the parent expression.